File: pset6/credit/credit.py
def main():
    CreditNumber = positiveNumber()
    if (check_isvalid(CreditNumber) == True):#Check if the number is valid
        cardBrand(CreditNumber) #Print the brand of the card
    else:
        print("INVALID\n")

#Function that ask the user for a positive number
def positiveNumber():
    while True:
        CreditNumber = input("Number: \n")
        n = int(CreditNumber)
        if n > 0:
            break
    return CreditNumber

#Function to check if the # of digits is correct to be a credit card
def check_isvalid(CreditNumber):
    validNumber = checksum(CreditNumber)
    lenght = len(CreditNumber)
    if ((lenght == 13 or lenght == 15 or lenght == 16) and (validNumber % 10 == 0)):
        return True
    else:
        return False

#Funtion to check Luhn’s Algorithm to determine if a creditcard number is valid
def checksum(CreditNumber):
    lenght = len(CreditNumber)
    ccn = int(CreditNumber)
    sum1 = 0
    sum2 = 0
    if lenght % 2 == 0:
        for i in range (lenght): #If the CreditCard is an even Number
            if lenght % 2 == 0:
                a = ccn % 10
                sum1 = sum1 + a
                ccn = int(ccn / 10)
                lenght -= 1
            else:
                b = 2 * (ccn % 10)
                bb = b // 10 + b % 10
                sum2 = sum2 + bb
                ccn = int(ccn /10)
                lenght -= 1
        return sum1 + sum2
    else:
        for i in range (lenght): #If the CreditCard is an odd Number
            if lenght % 2 == 0:
                a = 2 * (ccn % 10)
                aa = a //10 + a % 10
                sum1 = sum1 + aa
                ccn = int(ccn / 10)
                lenght -= 1
            else:
                b = ccn % 10
                sum2 = sum2 + b
                ccn = int(ccn / 10)
                lenght -= 1
        return sum1 + sum2

#Function to know what is the company's card
def cardBrand(ccn):
    if (len(ccn) == 15) and (ccn[:2] == '34') or (ccn[:2] == '37'):
        print("AMEX\n")
    elif (len(ccn) == 16) and (ccn[:2] == '51') or (ccn[:2] == '52') or (ccn[:2] == '53') or (ccn[:2] == '54') or (ccn[:2] == '55'):
        print("MASTERCARD\n")
    elif ((len(ccn) == 13 or len(ccn) == 16) and ccn[:1] == '4'):
        print("VISA\n")
    else:
        print("INVALID\n")


main()

File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>

void print_creditcard_brand(long);
bool checksum(long);
bool check_validity(long);
long get_positive_int(void);
int find_lenght(long);

int main(void)
{
    long creditcard_number = get_positive_int();
    if (check_validity(creditcard_number) == true)
    {
        print_creditcard_brand(creditcard_number);
    }
    else
    {
        printf("INVALID\n");
    }
}
//........................................
//Function that ask the user for a positive number
//........................................
long get_positive_int(void)
{
    long n;
    do
    {
        n = get_long("Credit Card Number: ");
    }
    while (n < 0);
    return n;
}
//........................................
//Function to check if the # of digits is correct to be a credit card
//........................................
bool check_validity(long creditcard_number)
{
    int len = find_lenght(creditcard_number);
    if ((len == 13 || len == 15 || len == 16) && checksum(creditcard_number))
    {
        return true;
    }
    else
    {
        return false;
    }
}
//........................................
//Function to know the lenght of a number
//........................................
int find_lenght(long n)
{
    int len;
    for (len = 0; n != 0; n /= 10)
    {
        len++;
    }
    return len;
}
//........................................
//Funtion to check Luhn’s Algorithm to determine if a creditcard number is valid
//........................................
bool checksum(long ccn)
{
    int sum = 0;
    for (int i = 0; ccn != 0; i++, ccn /= 10)
    {
        if (i % 2 == 0)
        {
            sum += ccn % 10;
        }
        else
        {
            int digit = 2 * (ccn % 10);
            sum += digit / 10 + digit % 10;
        }
    }
    return (sum % 10) == 0;
}
//........................................
// Function to know what is the company's card
//........................................
void print_creditcard_brand(long ccn)
{
    if ((ccn >= 34e13 && ccn < 35e13) || (ccn >= 37e13 && ccn < 38e13))
    {
        printf("AMEX\n");
    }
    else if (ccn >= 51e14 && ccn < 56e14)
    {
        printf("MASTERCARD\n");
    }
    else if ((ccn >= 4e12 && ccn < 5e12) || (ccn >= 4e15 && ccn < 5e15))
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}